src/Typedefs/Idris.idr

```diff
@@ -0,0 +1,260 @@
+module Typedefs.Idris
+
+import Data.Vect
+
+import public Typedefs.Names
+import public Typedefs.Typedefs
+
+%access public export
+
+||| Custom lookup for maps with dependent pairs
+custLookup : DecEq t => {w : t} -> {f, g : t -> Type}
+          -> List (v : t ** (f v, g v)) -> f w -> Eq (f w) => Maybe (g w)
+custLookup [] item = Nothing
+custLookup ((n ** (key, val)) :: xs) item {w} with (decEq n w)
+  -- if indices don't match then values can't possibly match since they have different type
+  custLookup ((n ** (key, val)) :: xs) item {w} | No _ = custLookup xs item
+  custLookup ((n ** (key, val)) :: xs) item {w=n} | Yes Refl =
+    if item == key then Just val -- index & key match, return value and its index
+                   else custLookup xs item -- index match but values differ
+
+TypeConstructor : Nat -> Type
+TypeConstructor Z = Type
+TypeConstructor (S n) = Type -> TypeConstructor n
+
+ApplyVect : Vect n Type -> TypeConstructor n -> Type
+ApplyVect [] c = c
+ApplyVect (x :: xs) c {n = S k} = ApplyVect xs (c x)
+
+||| A mapping from TDef to idris Type constructor
+SpecialList : Type
+SpecialList = List (n ** (TDefR n, TypeConstructor n))
+
+args : Vect k (String, TDef' (S n) a) -> TDef' (S n) a
+args []                 = T0
+args [(_,m)]            = m
+args ((_,m)::(_,l)::ms) = TSum (m :: l :: map snd ms)
+
+ReverseVect : (mkType : Vect n Type -> Type) -> TypeConstructor n
+ReverseVect mkType {n = Z} = mkType []
+ReverseVect mkType {n = (S k)} = \arg => ReverseVect (popHead mkType arg)
+  where
+    popHead : (Vect (S n) Type -> Type) -> Type -> Vect n Type -> Type
+    popHead f t ts = f (t :: ts)
+
+
+mutual
+  data Mu' : SpecialList -> Vect n Type -> TDef' (S n) a -> Type where
+    Inn : Ty' sp (Mu' sp tvars m :: tvars) m -> Mu' sp tvars m
+
+  Tnary' : SpecialList -> Vect n Type -> Vect (2 + k) (TDefR n) -> (Type -> Type -> Type) -> Type
```

Comment:
would it make sense to pull the `SpecialList` out into a
```idris
parameters (rep : SpecialList)
```
since I don't think it changes in any of the calls? (It's tempting to do the same for `Mu'`, but then I think it can't be implicit in the type of the constructor?)

Reply:
I tried to change it but it messes up with type inference for `Inn`. I'll leave it as is for now, since It will almost certainly change again once I implement `TermWrite` and `TermParse`